File: app/src/main/java/com/jaguar/toolkit/adapters/ToolsRecyclerAdapter.java
package com.jaguar.toolkit.adapters;

import android.app.AlertDialog;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.jaguar.toolkit.R;
import com.jaguar.toolkit.objects.Tool;

import java.util.List;

public class ToolsRecyclerAdapter extends RecyclerView.Adapter<ToolsRecyclerAdapter.ViewHolder> {
    Context context;
    List<Tool> toolsList;

    public ToolsRecyclerAdapter(Context context, List<Tool> toolsList) {
        this.context = context;
        this.toolsList = toolsList;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View ToolView = LayoutInflater.from(parent.getContext()).inflate(R.layout.card_tool, parent, false);
        return new ViewHolder(ToolView);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Tool tool = toolsList.get(position);
        holder.toolName.setText(tool.getName());
        holder.toolBtn.setImageResource(tool.getIcon());
        holder.setTool(tool);
    }

    @Override
    public int getItemCount() {
        return toolsList.size();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        final TextView toolName;
        final ImageButton toolBtn;
        public Tool tool;

        public ViewHolder(View itemView) {
            super(itemView);
            toolName = itemView.findViewById(R.id.tool_name);
            toolBtn = itemView.findViewById(R.id.tool_btn);
            toolBtn.setOnClickListener(v -> {
                AlertDialog.Builder builder = new AlertDialog.Builder(itemView.getContext());
                builder.setTitle(tool.getName()).setMessage("Selected the tool!").show();
            });
            toolBtn.setOnLongClickListener(v -> {
                Toast.makeText(itemView.getContext(), tool.getDescription(), Toast.LENGTH_SHORT).show();
                return true;
            });
        }

        public void setTool(Tool tool) {
            this.tool = tool;
        }
    }
}

File: app/src/main/java/com/jaguar/toolkit/adapters/ToolsRecyclerAdapter.kt
package com.jaguar.toolkit.adapters

import android.content.Context
import android.content.Intent
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.TextView
import android.widget.Toast
import androidx.recyclerview.widget.RecyclerView
import com.jaguar.toolkit.R
import com.jaguar.toolkit.components.ToolCard
import com.jaguar.toolkit.objects.Tool

class ToolsRecyclerAdapter(context: Context?, private val toolsList: List<Tool<*>>) :
    RecyclerView.Adapter<ToolsRecyclerAdapter.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val toolCard = ToolCard(parent.context).build()
        return ViewHolder(toolCard)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val tool = toolsList[position]
        holder.bind(tool)
    }

    override fun getItemCount(): Int = toolsList.size

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val toolName: TextView = itemView.findViewById(R.id.tool_name)
        private val toolBtn: ImageButton = itemView.findViewById(R.id.tool_btn)
        private var tool: Tool<*>? = null

        init {
            toolBtn.setOnClickListener {
                tool?.let {
                    val intent = Intent(itemView.context, it.getActivity())
                    itemView.context.startActivity(intent)
                }
            }
            toolBtn.setOnLongClickListener {
                tool?.let {
                    Toast.makeText(itemView.context, it.getDescription(), Toast.LENGTH_SHORT).show()
                }
                true
            }
        }

        fun bind(tool: Tool<*>) {
            this.tool = tool
            toolName.text = tool.getName()
            toolBtn.setImageResource(tool.getIcon())
        }
    }
}